src/Subscription.js

```diff
@@ -99,41 +86,56 @@ export default class Subscription extends EventEmitter {
     }
 
     /**
-     * Gap check: If the msg contains the previousOffset, and we know the lastReceivedOffset,
-     * and the previousOffset is larger than what has been received, we have a gap!
+     * Gap check: If the msg contains the previousMsgRef, and we know the lastReceivedMsgRef,
+     * and the previousMsgRef is larger than what has been received, we have a gap!
      */
-    checkForGap(previousOffset) {
-        return previousOffset != null &&
-            this.lastReceivedOffset != null &&
-            previousOffset > this.lastReceivedOffset
+    checkForGap(previousMsgRef, key) {
+        return previousMsgRef != null &&
+            this.lastReceivedMsgRef[key] !== undefined &&
+            Subscription.compareMessageRefs(previousMsgRef, this.lastReceivedMsgRef[key]) === 1
     }
 
     handleMessage(msg, isResend = false) {
-        if (msg.previousOffset == null) {
+        if (msg.version !== 30) {
+            throw new Error(`Can handle only StreamMessageV30, not version ${msg.version}`)
+        }
+        if (msg.prevMsgRef == null) {
             debug('handleMessage: prevOffset is null, gap detection is impossible! message: %o', msg)
         }
 
+        const key = msg.getPublisherId() + msg.messageId.msgChainId
+
         // TODO: check this.options.resend_last ?
-        // If resending, queue broadcasted messages
+        // If resending, queue broadcast messages
         if (this.resending && !isResend) {
             this.queue.push(msg)
-        } else if (this.checkForGap(msg.previousOffset) && !this.resending) {
+        } else if (this.checkForGap(msg.prevMsgRef, key) && !this.resending) {
             // Queue the message to be processed after resend
             this.queue.push(msg)
 
-            const from = this.lastReceivedOffset + 1
-            const to = msg.previousOffset
-            debug('Gap detected, requesting resend for stream %s from %d to %d', this.streamId, from, to)
-            this.emit('gap', from, to)
-        } else if (this.lastReceivedOffset != null && msg.offset <= this.lastReceivedOffset) {
-            // Prevent double-processing of messages for any reason
-            debug('Sub %s already received message: %d, lastReceivedOffset: %d. Ignoring message.', this.id, msg.offset, this.lastReceivedOffset)
+            const from = this.lastReceivedMsgRef[key] // cannot know the first missing message so there will be a duplicate received
```

Comment:
I think it's better to do it in a separate PR so that we have more or less one PR per feature and per JIRA issue. As soon as this is merged, I could tackle it.